pkg/controller/controlplane/valuesprovider.go

```diff
@@ -587,6 +625,74 @@ func getCRCChartValues(
 	return values, nil
 }
 
+// getIPAMChartValues collects and returns the ipam-controller chart values.
+func getIPAMChartValues(
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+	checksums map[string]string,
+	scaledDown bool,
+) (map[string]interface{}, error) {
+
+	mode := "ipv4"
+	if cluster.Shoot.Spec.Networking != nil &&
+		len(cluster.Shoot.Spec.Networking.IPFamilies) == 1 &&
+		cluster.Shoot.Spec.Networking.IPFamilies[0] == v1beta1.IPFamilyIPv6 {
+		mode = "ipv6"
+	}
+	if cluster.Shoot.Spec.Networking != nil &&
+		len(cluster.Shoot.Spec.Networking.IPFamilies) == 2 {
+		mode = "dual-stack"
+	}
+
+	nodeCidrMaskSizeIPv4 := int32(24)
+	nodeCidrMaskSizeIPv6 := int32(64)
+	if cluster.Shoot.Spec.Kubernetes.KubeControllerManager != nil && cluster.Shoot.Spec.Kubernetes.KubeControllerManager.NodeCIDRMaskSize != nil {
+		if cluster.Shoot.Spec.Networking != nil &&
+			len(cluster.Shoot.Spec.Networking.IPFamilies) == 1 &&
+			cluster.Shoot.Spec.Networking.IPFamilies[0] == v1beta1.IPFamilyIPv4 {
+			nodeCidrMaskSizeIPv4 = *cluster.Shoot.Spec.Kubernetes.KubeControllerManager.NodeCIDRMaskSize
+		}
+		if cluster.Shoot.Spec.Networking != nil &&
+			len(cluster.Shoot.Spec.Networking.IPFamilies) == 1 &&
+			cluster.Shoot.Spec.Networking.IPFamilies[0] == v1beta1.IPFamilyIPv6 {
+			nodeCidrMaskSizeIPv6 = *cluster.Shoot.Spec.Kubernetes.KubeControllerManager.NodeCIDRMaskSize
+		}
+	}
+
+	podNetwork := "192.168.0.0/16"
+	if slices.Contains(cluster.Shoot.Spec.Networking.IPFamilies, v1beta1.IPFamilyIPv4) {
+		podNetwork = extensionscontroller.GetPodNetwork(cluster)
+	}
+
+	values := map[string]interface{}{
+		"enabled":     true,
+		"replicas":    extensionscontroller.GetControlPlaneReplicas(cluster, scaledDown, 1),
+		"clusterName": cp.Namespace,
+		"podNetwork":  podNetwork,
+		"podAnnotations": map[string]interface{}{
+			"checksum/secret-cloudprovider": checksums[v1beta1constants.SecretNameCloudProvider],
+		},
+		"podLabels": map[string]interface{}{
+			v1beta1constants.LabelPodMaintenanceRestart: "true",
+		},
+		"region":               cp.Spec.Region,
+		"mode":                 mode,
+		"nodeCIDRMaskSizeIPv4": nodeCidrMaskSizeIPv4,
+		"nodeCIDRMaskSizeIPv6": nodeCidrMaskSizeIPv6,
+	}
+	enabled := (cluster.Shoot.Spec.Networking != nil &&
```

Comment:
1. I feel that this condition is used in many places already, hence it would be nice if we had an utility function for it already.
2. Maybe a clarification for the enablement and the FeatureGate. My understanding is the following: we need the IPAM controller no matter what for dual stack and ipv6 and hiding behind the feature gate is not changing much. I though the FG was there to enable ipv4 using the new controller for all shoots (including the ipv4 only shoots)

Reply:
Done and added additional test case to test all feature gate scenarios.